File: clients/scala/src/main/scala/cloud/fiducia/Fiducia.scala
// Fiducia HTTP client (Scala, JDK 11+). Transport: java.net.http.HttpClient.
// JSON: com.lihaoyi ujson (the only dependency). Implements PROTOCOL.md.
//
//   val c = new cloud.fiducia.FiduciaClient("https://api.fiducia.cloud")
//   val grant = c.lockAcquire("orders/checkout", ttlMs = Some(30000L))
//   val token = grant("result")("output")("fencing_token").num.toLong
//   c.lockRelease("orders/checkout", "worker-a", token)

package cloud.fiducia

import java.net.URI
import java.net.URLEncoder
import java.net.http.{HttpClient, HttpRequest, HttpResponse}
import java.nio.charset.StandardCharsets
import java.time.Duration
import scala.language.implicitConversions
import scala.util.control.NonFatal

/** Raised on any HTTP response with status >= 300. `body` is the parsed JSON
  * response (`ujson.Null` when the response body was empty). */
final case class FiduciaException(status: Int, body: ujson.Value)
    extends RuntimeException(s"fiducia: HTTP $status")

/** A thin, dependency-light HTTP wrapper over the fiducia.cloud contract. Every
  * method returns the parsed JSON response as a [[ujson.Value]] and throws a
  * [[FiduciaException]] on any non-2xx status.
  *
  * Optional arguments are expressed as `Option`s: a field is included in the
  * request body only when the caller passes `Some(..)` (this matters for the
  * compare-and-swap semantics of `prevRevision`, `holder`, etc.). A supplied
  * `Some(0)` — e.g. `prevRevision = Some(0L)`, meaning "must not exist" — is
  * always sent; it is not dropped as a falsy value.
  *
  * '''Numeric precision:''' ujson's `Num` is `Double`-backed, so 64-bit integers
  * (fencing tokens, KV revisions) are exact only up to 2^53. This client returns
  * the parsed response tree untouched — it never narrows a token to a different
  * type — so read such values straight off the tree, e.g.
  * `resp("result")("output")("fencing_token").num.toLong`. On the request side a
  * `Long` is likewise carried through a `Num`; integral values render as plain
  * integers on the wire (no `.0`, no exponent), exactly up to 2^53 and rounded
  * beyond it. Real monotonic counters never approach 2^53; this ceiling is an
  * inherent limit of the ujson AST, not of the wire protocol.
  *
  * @param baseUrl        service base URL; a trailing slash is trimmed
  * @param requestTimeout connect + per-request timeout applied to every call;
  *                       defaults to 30s. Pass `None` to disable it (e.g. for a
  *                       caller that deliberately blocks on a long `wait`).
  */
class FiduciaClient(
    baseUrl: String,
    requestTimeout: Option[Duration] = Some(Duration.ofSeconds(30))
) {

  private val base: String = baseUrl.replaceAll("/+$", "")
  // Redirects are NOT followed. A 3xx on a mutating POST/PUT/DELETE must surface
  // as an error (status >= 300) rather than silently re-submitting the operation
  // to the Location and duplicating a lock grant / FIFO queue slot. NEVER is also
  // the java.net.http default; pinned here so it cannot drift.
  private val http: HttpClient = {
    val b = HttpClient.newBuilder().followRedirects(HttpClient.Redirect.NEVER)
    requestTimeout.foreach(t => b.connectTimeout(t))
    b.build()
  }

  // --- misc ----------------------------------------------------------------
  def health(): ujson.Value = request("GET", "/healthz")
  def status(): ujson.Value = request("GET", "/v1/status")

  // --- locks ---------------------------------------------------------------
  def lockGet(key: String): ujson.Value =
    request("GET", s"/v1/locks?key=${enc(key)}")

  def lockAcquire(
      key: String,
      holder: Option[String] = None,
      ttlMs: Option[Long] = None,
      wait: Boolean = true
  ): ujson.Value = {
    val b = ujson.Obj("key" -> ujson.Str(key), "wait" -> ujson.Bool(wait))
    putOpt(b, "holder", holder.map(h => ujson.Str(h)))
    putOpt(b, "ttl_ms", ttlMs.map(t => ujson.Num(t.toDouble)))
    request("POST", "/v1/locks/acquire", Some(b))
  }

  /** Atomic multi-key UNION lock: all-or-nothing across the whole set. */
  def lockAcquireMany(
      keys: Seq[String],
      holder: Option[String] = None,
      ttlMs: Option[Long] = None,
      wait: Boolean = true
  ): ujson.Value = {
    val b = ujson.Obj(
      "keys" -> ujson.Arr(keys.map(k => ujson.Str(k)): _*),
      "wait" -> ujson.Bool(wait)
    )
    putOpt(b, "holder", holder.map(h => ujson.Str(h)))
    putOpt(b, "ttl_ms", ttlMs.map(t => ujson.Num(t.toDouble)))
    request("POST", "/v1/locks/acquire", Some(b))
  }

  def tryLock(key: String, holder: Option[String] = None, ttlMs: Option[Long] = None): ujson.Value =
    lockAcquire(key, holder, ttlMs, wait = false)

  def mustLock(key: String, holder: Option[String] = None, ttlMs: Option[Long] = None): ujson.Value =
    lockAcquire(key, holder, ttlMs, wait = true)

  /** Alias for [[mustLock]]. */
  def lock(key: String, holder: Option[String] = None, ttlMs: Option[Long] = None): ujson.Value =
    mustLock(key, holder, ttlMs)

  /** `key` is accepted for symmetry but is not sent (release is by token). */
  def lockRelease(key: String, holder: String, fencingToken: Long): ujson.Value =
    request("POST", "/v1/locks/release", Some(ujson.Obj(
      "holder" -> ujson.Str(holder),
      "fencing_token" -> ujson.Num(fencingToken.toDouble)
    )))

  // --- semaphores ----------------------------------------------------------
  def semaphoreGet(key: String): ujson.Value =
    request("GET", s"/v1/semaphores?key=${enc(key)}")

  def semaphoreAcquire(
      key: String,
      limit: Long,
      holder: Option[String] = None,
      ttlMs: Option[Long] = None,
      wait: Boolean = true
  ): ujson.Value = {
    val b = ujson.Obj(
      "key" -> ujson.Str(key),
      "limit" -> ujson.Num(limit.toDouble),
      "wait" -> ujson.Bool(wait)
    )
    putOpt(b, "holder", holder.map(h => ujson.Str(h)))
    putOpt(b, "ttl_ms", ttlMs.map(t => ujson.Num(t.toDouble)))
    request("POST", "/v1/semaphores/acquire", Some(b))
  }

  def trySemaphore(key: String, limit: Long, holder: Option[String] = None, ttlMs: Option[Long] = None): ujson.Value =
    semaphoreAcquire(key, limit, holder, ttlMs, wait = false)

  def mustSemaphore(key: String, limit: Long, holder: Option[String] = None, ttlMs: Option[Long] = None): ujson.Value =
    semaphoreAcquire(key, limit, holder, ttlMs, wait = true)

  /** Alias for [[mustSemaphore]]. */
  def semaphore(key: String, limit: Long, holder: Option[String] = None, ttlMs: Option[Long] = None): ujson.Value =
    mustSemaphore(key, limit, holder, ttlMs)

  def semaphoreRelease(key: String, holder: String, fencingToken: Long): ujson.Value =
    request("POST", "/v1/semaphores/release", Some(ujson.Obj(
      "key" -> ujson.Str(key),
      "holder" -> ujson.Str(holder),
      "fencing_token" -> ujson.Num(fencingToken.toDouble)
    )))

  // --- idempotency ---------------------------------------------------------
  def idempotencyGet(key: String): ujson.Value =
    request("GET", s"/v1/idempotency?key=${enc(key)}")

  def idempotencyClaim(
      key: String,
      owner: Option[String] = None,
      ttlMs: Option[Long] = None,
      ttl: Option[String] = None,
      metadata: Option[ujson.Value] = None
  ): ujson.Value = {
    val b = ujson.Obj("key" -> ujson.Str(key))
    putOpt(b, "owner", owner.map(o => ujson.Str(o)))
    putOpt(b, "ttl_ms", ttlMs.map(t => ujson.Num(t.toDouble)))
    putOpt(b, "ttl", ttl.map(t => ujson.Str(t)))
    putOpt(b, "metadata", metadata)
    request("POST", "/v1/idempotency/claim", Some(b))
  }

  def idempotencyComplete(
      key: String,
      owner: String,
      fencingToken: Long,
      result: Option[ujson.Value] = None
  ): ujson.Value = {
    val b = ujson.Obj(
      "key" -> ujson.Str(key),
      "owner" -> ujson.Str(owner),
      "fencing_token" -> ujson.Num(fencingToken.toDouble)
    )
    putOpt(b, "result", result)
    request("POST", "/v1/idempotency/complete", Some(b))
  }

  // --- reader-writer locks -------------------------------------------------
  def rwAcquireRead(key: String, ttlMs: Option[Long] = None, wait: Boolean = true): ujson.Value = {
    val b = ujson.Obj("wait" -> ujson.Bool(wait))
    putOpt(b, "ttl_ms", ttlMs.map(t => ujson.Num(t.toDouble)))
    request("POST", s"/v1/rw/${enc(key)}/read", Some(b))
  }

  def rwEndRead(key: String, lockId: String): ujson.Value =
    request("POST", s"/v1/rw/${enc(key)}/read/end", Some(ujson.Obj("lock_id" -> ujson.Str(lockId))))

  def rwAcquireWrite(key: String, ttlMs: Option[Long] = None, wait: Boolean = true): ujson.Value = {
    val b = ujson.Obj("wait" -> ujson.Bool(wait))
    putOpt(b, "ttl_ms", ttlMs.map(t => ujson.Num(t.toDouble)))
    request("POST", s"/v1/rw/${enc(key)}/write", Some(b))
  }

  def rwEndWrite(key: String, lockId: String): ujson.Value =
    request("POST", s"/v1/rw/${enc(key)}/write/end", Some(ujson.Obj("lock_id" -> ujson.Str(lockId))))

  // --- config KV -----------------------------------------------------------
  def kvGet(key: String): ujson.Value =
    request("GET", s"/v1/kv?key=${enc(key)}")

  /** `prevRevision` is a compare-and-swap guard (`Some(0)` = must-not-exist). */
  def kvPut(
      key: String,
      value: ujson.Value,
      ttlMs: Option[Long] = None,
      prevRevision: Option[Long] = None
  ): ujson.Value = {
    val b = ujson.Obj("value" -> value)
    putOpt(b, "ttl_ms", ttlMs.map(t => ujson.Num(t.toDouble)))
    putOpt(b, "prev_revision", prevRevision.map(r => ujson.Num(r.toDouble)))
    request("PUT", s"/v1/kv?key=${enc(key)}", Some(b))
  }

  def kvDelete(key: String): ujson.Value =
    request("DELETE", s"/v1/kv?key=${enc(key)}")

  def kvList(prefix: String): ujson.Value =
    request("GET", s"/v1/kv?prefix=${enc(prefix)}")

  // --- rate limiting -------------------------------------------------------
  def rateLimitGet(tenant: String, key: String): ujson.Value =
    request("GET", s"/v1/rate-limit/${enc(tenant)}/${enc(key)}")

  def rateLimitCheck(
      tenant: String,
      key: String,
      algorithm: String,
      limit: Long,
      windowMs: Long,
      refillPerSecond: Option[Double] = None,
      cost: Option[Long] = None
  ): ujson.Value = {
    val b = ujson.Obj(
      "algorithm" -> ujson.Str(algorithm),
      "limit" -> ujson.Num(limit.toDouble),
      "window_ms" -> ujson.Num(windowMs.toDouble)
    )
    putOpt(b, "refill_per_second", refillPerSecond.map(r => ujson.Num(r)))
    putOpt(b, "cost", cost.map(c => ujson.Num(c.toDouble)))
    request("POST", s"/v1/rate-limit/${enc(tenant)}/${enc(key)}/check", Some(b))
  }

  // --- cron & scheduling ---------------------------------------------------
  def scheduleGet(name: String): ujson.Value =
    request("GET", s"/v1/cron/schedules/${enc(name)}")

  /** `target` is an arbitrary JSON object, e.g. `ujson.Obj("kind" -> "webhook", "url" -> "..")`. */
  def scheduleUpsert(
      name: String,
      target: ujson.Value,
      cron: Option[String] = None,
      oneShotAtMs: Option[Long] = None,
      delivery: Option[String] = None,
      maxRetries: Option[Long] = None
  ): ujson.Value = {
    val b = ujson.Obj("target" -> target)
    putOpt(b, "cron", cron.map(c => ujson.Str(c)))
    putOpt(b, "one_shot_at_ms", oneShotAtMs.map(t => ujson.Num(t.toDouble)))
    putOpt(b, "delivery", delivery.map(d => ujson.Str(d)))
    putOpt(b, "max_retries", maxRetries.map(m => ujson.Num(m.toDouble)))
    request("PUT", s"/v1/cron/schedules/${enc(name)}", Some(b))
  }

  def scheduleRecordRun(name: String, fireId: String, firedAtMs: Option[Long] = None): ujson.Value = {
    val b = ujson.Obj("fire_id" -> ujson.Str(fireId))
    putOpt(b, "fired_at_ms", firedAtMs.map(t => ujson.Num(t.toDouble)))
    request("POST", s"/v1/cron/schedules/${enc(name)}/runs", Some(b))
  }

  def scheduleHistory(name: String): ujson.Value =
    request("GET", s"/v1/cron/schedules/${enc(name)}/history")

  // --- leader election -----------------------------------------------------
  def electionGet(name: String): ujson.Value =
    request("GET", s"/v1/elections/${enc(name)}")

  def electionCampaign(
      name: String,
      candidate: String,
      ttlMs: Long,
      metadata: Option[ujson.Value] = None
  ): ujson.Value = {
    val b = ujson.Obj("candidate" -> ujson.Str(candidate), "ttl_ms" -> ujson.Num(ttlMs.toDouble))
    putOpt(b, "metadata", metadata)
    request("POST", s"/v1/elections/${enc(name)}/campaign", Some(b))
  }

  def electionRenew(name: String, candidate: String, fencingToken: Long): ujson.Value =
    request("POST", s"/v1/elections/${enc(name)}/renew", Some(ujson.Obj(
      "candidate" -> ujson.Str(candidate),
      "fencing_token" -> ujson.Num(fencingToken.toDouble)
    )))

  def electionResign(name: String, candidate: String, fencingToken: Long): ujson.Value =
    request("POST", s"/v1/elections/${enc(name)}/resign", Some(ujson.Obj(
      "candidate" -> ujson.Str(candidate),
      "fencing_token" -> ujson.Num(fencingToken.toDouble)
    )))

  // --- service discovery ---------------------------------------------------
  def serviceInstances(service: String): ujson.Value =
    request("GET", s"/v1/services/${enc(service)}")

  def serviceRegister(
      service: String,
      instanceId: String,
      address: String,
      ttlMs: Long,
      metadata: Option[ujson.Value] = None
  ): ujson.Value = {
    val b = ujson.Obj("address" -> ujson.Str(address), "ttl_ms" -> ujson.Num(ttlMs.toDouble))
    putOpt(b, "metadata", metadata)
    request("PUT", s"/v1/services/${enc(service)}/instances/${enc(instanceId)}", Some(b))
  }

  def serviceHeartbeat(service: String, instanceId: String, ttlMs: Option[Long] = None): ujson.Value = {
    val b = ujson.Obj()
    putOpt(b, "ttl_ms", ttlMs.map(t => ujson.Num(t.toDouble)))
    request("POST", s"/v1/services/${enc(service)}/instances/${enc(instanceId)}/heartbeat", Some(b))
  }

  def serviceDeregister(service: String, instanceId: String): ujson.Value =
    request("DELETE", s"/v1/services/${enc(service)}/instances/${enc(instanceId)}")

  def serviceList(): ujson.Value =
    request("GET", "/v1/services")

  // --- internals -----------------------------------------------------------
  private def putOpt(b: ujson.Obj, key: String, value: Option[ujson.Value]): Unit =
    value.foreach(v => b.obj(key) = v)

  private def enc(s: String): String =
    URLEncoder.encode(s, StandardCharsets.UTF_8).replace("+", "%20")

  private def request(method: String, path: String, body: Option[ujson.Value] = None): ujson.Value = {
    val builder = HttpRequest.newBuilder(URI.create(base + path))
    requestTimeout.foreach(t => builder.timeout(t))
    body match {
      case Some(b) =>
        builder.header("content-type", "application/json")
        builder.method(method, HttpRequest.BodyPublishers.ofString(ujson.write(b)))
      case None =>
        builder.method(method, HttpRequest.BodyPublishers.noBody())
    }
    val response = http.send(builder.build(), HttpResponse.BodyHandlers.ofString())
    val text = response.body()
    // An empty/204 body is null; a non-JSON body (e.g. a proxy's plain-text 502)
    // must not crash the parser and lose the status — fall back to the raw text.
    val data: ujson.Value =
      if (text == null || text.isEmpty) ujson.Null
      else
        try ujson.read(text)
        catch { case NonFatal(_) => ujson.Str(text) }
    if (response.statusCode() >= 300) throw FiduciaException(response.statusCode(), data)
    data
  }
}

File: clients/swift/Sources/Fiducia/Fiducia.swift
// Fiducia HTTP client (Swift). Zero-dependency — Foundation URLSession + JSONSerialization.
// Implements PROTOCOL.md.
//
//   let c = FiduciaClient(baseURL: "https://api.fiducia.cloud")
//   let grant = try await c.lockAcquire("orders/checkout", ttlMs: 30000)
//   let out = ((grant as? [String: Any])?["result"] as? [String: Any])?["output"] as? [String: Any]
//   _ = try await c.lockRelease("orders/checkout", holder: "worker-a", fencingToken: out?["fencing_token"] ?? 0)

import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

/// A non-2xx response. Carries the numeric HTTP status and the parsed JSON body
/// (the natural `Any` produced by `JSONSerialization`, or `nil` for an empty body).
public struct FiduciaError: Error, CustomStringConvertible {
    public let status: Int
    public let body: Any?

    public init(status: Int, body: Any?) {
        self.status = status
        self.body = body
    }

    public var description: String { "fiducia: HTTP \(status)" }
}

/// Thin async HTTP wrapper over the fiducia.cloud contract.
///
/// Every method returns the parsed JSON response as `Any` (a `[String: Any]`,
/// `[Any]`, `NSNumber`, `String`, or `NSNull`). An empty response body comes back
/// as `NSNull()`. On HTTP status >= 300 a `FiduciaError` is thrown.
public final class FiduciaClient {
    private let baseURL: String
    private let session: URLSession

    /// Optional per-request timeout (seconds) applied to every request.
    public var requestTimeout: TimeInterval?

    public init(baseURL: String, session: URLSession = .shared, requestTimeout: TimeInterval? = nil) {
        var trimmed = baseURL
        while trimmed.hasSuffix("/") { trimmed.removeLast() }
        self.baseURL = trimmed
        self.session = session
        self.requestTimeout = requestTimeout
    }

    // MARK: - misc

    public func health() async throws -> Any { try await request("GET", "/healthz") }
    public func status() async throws -> Any { try await request("GET", "/v1/status") }

    // MARK: - locks

    public func lockGet(_ key: String) async throws -> Any {
        try await request("GET", "/v1/locks?key=\(Self.enc(key))")
    }

    public func lockAcquire(_ key: String, holder: String? = nil, ttlMs: Int? = nil, wait: Bool = true) async throws -> Any {
        var body: [String: Any] = ["key": key, "wait": wait]
        if let holder = holder { body["holder"] = holder }
        if let ttlMs = ttlMs { body["ttl_ms"] = ttlMs }
        return try await request("POST", "/v1/locks/acquire", body: body)
    }

    /// Atomically acquire a union lock over several keys.
    public func lockAcquireMany(_ keys: [String], holder: String? = nil, ttlMs: Int? = nil, wait: Bool = true) async throws -> Any {
        var body: [String: Any] = ["keys": keys, "wait": wait]
        if let holder = holder { body["holder"] = holder }
        if let ttlMs = ttlMs { body["ttl_ms"] = ttlMs }
        return try await request("POST", "/v1/locks/acquire", body: body)
    }

    public func tryLock(_ key: String, holder: String? = nil, ttlMs: Int? = nil) async throws -> Any {
        try await lockAcquire(key, holder: holder, ttlMs: ttlMs, wait: false)
    }

    public func mustLock(_ key: String, holder: String? = nil, ttlMs: Int? = nil) async throws -> Any {
        try await lockAcquire(key, holder: holder, ttlMs: ttlMs, wait: true)
    }

    /// Alias for `mustLock`.
    public func lock(_ key: String, holder: String? = nil, ttlMs: Int? = nil) async throws -> Any {
        try await mustLock(key, holder: holder, ttlMs: ttlMs)
    }

    /// `key` is accepted for call-site symmetry but is intentionally omitted from the body.
    public func lockRelease(_ key: String, holder: String, fencingToken: Any) async throws -> Any {
        try await request("POST", "/v1/locks/release", body: ["holder": holder, "fencing_token": fencingToken])
    }

    // MARK: - semaphores

    public func semaphoreGet(_ key: String) async throws -> Any {
        try await request("GET", "/v1/semaphores?key=\(Self.enc(key))")
    }

    public func semaphoreAcquire(_ key: String, limit: Int, holder: String? = nil, ttlMs: Int? = nil, wait: Bool = true) async throws -> Any {
        var body: [String: Any] = ["key": key, "limit": limit, "wait": wait]
        if let holder = holder { body["holder"] = holder }
        if let ttlMs = ttlMs { body["ttl_ms"] = ttlMs }
        return try await request("POST", "/v1/semaphores/acquire", body: body)
    }

    public func trySemaphore(_ key: String, limit: Int, holder: String? = nil, ttlMs: Int? = nil) async throws -> Any {
        try await semaphoreAcquire(key, limit: limit, holder: holder, ttlMs: ttlMs, wait: false)
    }

    public func mustSemaphore(_ key: String, limit: Int, holder: String? = nil, ttlMs: Int? = nil) async throws -> Any {
        try await semaphoreAcquire(key, limit: limit, holder: holder, ttlMs: ttlMs, wait: true)
    }

    /// Alias for `mustSemaphore`.
    public func semaphore(_ key: String, limit: Int, holder: String? = nil, ttlMs: Int? = nil) async throws -> Any {
        try await mustSemaphore(key, limit: limit, holder: holder, ttlMs: ttlMs)
    }

    public func semaphoreRelease(_ key: String, holder: String, fencingToken: Any) async throws -> Any {
        try await request("POST", "/v1/semaphores/release", body: ["key": key, "holder": holder, "fencing_token": fencingToken])
    }

    // MARK: - idempotency keys

    public func idempotencyGet(_ key: String) async throws -> Any {
        try await request("GET", "/v1/idempotency?key=\(Self.enc(key))")
    }

    public func idempotencyClaim(_ key: String, owner: String? = nil, ttlMs: Int? = nil, ttl: String? = nil, metadata: [String: Any]? = nil) async throws -> Any {
        var body: [String: Any] = ["key": key]
        if let owner = owner { body["owner"] = owner }
        if let ttlMs = ttlMs { body["ttl_ms"] = ttlMs }
        if let ttl = ttl { body["ttl"] = ttl }
        if let metadata = metadata { body["metadata"] = metadata }
        return try await request("POST", "/v1/idempotency/claim", body: body)
    }

    public func idempotencyComplete(_ key: String, owner: String, fencingToken: Any, result: [String: Any]? = nil) async throws -> Any {
        var body: [String: Any] = ["key": key, "owner": owner, "fencing_token": fencingToken]
        if let result = result { body["result"] = result }
        return try await request("POST", "/v1/idempotency/complete", body: body)
    }

    // MARK: - reader-writer locks

    public func rwAcquireRead(_ key: String, ttlMs: Int? = nil, wait: Bool = true) async throws -> Any {
        var body: [String: Any] = ["wait": wait]
        if let ttlMs = ttlMs { body["ttl_ms"] = ttlMs }
        return try await request("POST", "/v1/rw/\(Self.enc(key))/read", body: body)
    }

    public func rwEndRead(_ key: String, lockId: String) async throws -> Any {
        try await request("POST", "/v1/rw/\(Self.enc(key))/read/end", body: ["lock_id": lockId])
    }

    public func rwAcquireWrite(_ key: String, ttlMs: Int? = nil, wait: Bool = true) async throws -> Any {
        var body: [String: Any] = ["wait": wait]
        if let ttlMs = ttlMs { body["ttl_ms"] = ttlMs }
        return try await request("POST", "/v1/rw/\(Self.enc(key))/write", body: body)
    }

    public func rwEndWrite(_ key: String, lockId: String) async throws -> Any {
        try await request("POST", "/v1/rw/\(Self.enc(key))/write/end", body: ["lock_id": lockId])
    }

    // MARK: - config KV

    public func kvGet(_ key: String) async throws -> Any {
        try await request("GET", "/v1/kv?key=\(Self.enc(key))")
    }

    public func kvPut(_ key: String, value: Any, ttlMs: Int? = nil, prevRevision: Int? = nil) async throws -> Any {
        var body: [String: Any] = ["value": value]
        if let ttlMs = ttlMs { body["ttl_ms"] = ttlMs }
        if let prevRevision = prevRevision { body["prev_revision"] = prevRevision }
        return try await request("PUT", "/v1/kv?key=\(Self.enc(key))", body: body)
    }

    public func kvDelete(_ key: String) async throws -> Any {
        try await request("DELETE", "/v1/kv?key=\(Self.enc(key))")
    }

    public func kvList(prefix: String) async throws -> Any {
        try await request("GET", "/v1/kv?prefix=\(Self.enc(prefix))")
    }

    // MARK: - rate limiting

    public func rateLimitGet(tenant: String, key: String) async throws -> Any {
        try await request("GET", "/v1/rate-limit/\(Self.enc(tenant))/\(Self.enc(key))")
    }

    public func rateLimitCheck(tenant: String, key: String, algorithm: String, limit: Int, windowMs: Int, refillPerSecond: Double? = nil, cost: Int? = nil) async throws -> Any {
        var body: [String: Any] = ["algorithm": algorithm, "limit": limit, "window_ms": windowMs]
        if let refillPerSecond = refillPerSecond { body["refill_per_second"] = refillPerSecond }
        if let cost = cost { body["cost"] = cost }
        return try await request("POST", "/v1/rate-limit/\(Self.enc(tenant))/\(Self.enc(key))/check", body: body)
    }

    // MARK: - cron & scheduling

    public func scheduleGet(_ name: String) async throws -> Any {
        try await request("GET", "/v1/cron/schedules/\(Self.enc(name))")
    }

    public func scheduleUpsert(_ name: String, target: [String: Any], cron: String? = nil, oneShotAtMs: Int? = nil, delivery: String? = nil, maxRetries: Int? = nil) async throws -> Any {
        var body: [String: Any] = ["target": target]
        if let cron = cron { body["cron"] = cron }
        if let oneShotAtMs = oneShotAtMs { body["one_shot_at_ms"] = oneShotAtMs }
        if let delivery = delivery { body["delivery"] = delivery }
        if let maxRetries = maxRetries { body["max_retries"] = maxRetries }
        return try await request("PUT", "/v1/cron/schedules/\(Self.enc(name))", body: body)
    }

    public func scheduleRecordRun(_ name: String, fireId: String, firedAtMs: Int? = nil) async throws -> Any {
        var body: [String: Any] = ["fire_id": fireId]
        if let firedAtMs = firedAtMs { body["fired_at_ms"] = firedAtMs }
        return try await request("POST", "/v1/cron/schedules/\(Self.enc(name))/runs", body: body)
    }

    public func scheduleHistory(_ name: String) async throws -> Any {
        try await request("GET", "/v1/cron/schedules/\(Self.enc(name))/history")
    }

    // MARK: - leader election

    public func electionGet(_ name: String) async throws -> Any {
        try await request("GET", "/v1/elections/\(Self.enc(name))")
    }

    public func electionCampaign(_ name: String, candidate: String, ttlMs: Int, metadata: [String: Any]? = nil) async throws -> Any {
        var body: [String: Any] = ["candidate": candidate, "ttl_ms": ttlMs]
        if let metadata = metadata { body["metadata"] = metadata }
        return try await request("POST", "/v1/elections/\(Self.enc(name))/campaign", body: body)
    }

    public func electionRenew(_ name: String, candidate: String, fencingToken: Any) async throws -> Any {
        try await request("POST", "/v1/elections/\(Self.enc(name))/renew", body: ["candidate": candidate, "fencing_token": fencingToken])
    }

    public func electionResign(_ name: String, candidate: String, fencingToken: Any) async throws -> Any {
        try await request("POST", "/v1/elections/\(Self.enc(name))/resign", body: ["candidate": candidate, "fencing_token": fencingToken])
    }

    // MARK: - service discovery

    public func serviceList() async throws -> Any {
        try await request("GET", "/v1/services")
    }

    public func serviceInstances(_ service: String) async throws -> Any {
        try await request("GET", "/v1/services/\(Self.enc(service))")
    }

    public func serviceRegister(_ service: String, instanceId: String, address: String, ttlMs: Int, metadata: [String: Any]? = nil) async throws -> Any {
        var body: [String: Any] = ["address": address, "ttl_ms": ttlMs]
        if let metadata = metadata { body["metadata"] = metadata }
        return try await request("PUT", "/v1/services/\(Self.enc(service))/instances/\(Self.enc(instanceId))", body: body)
    }

    public func serviceHeartbeat(_ service: String, instanceId: String, ttlMs: Int? = nil) async throws -> Any {
        var body: [String: Any] = [:]
        if let ttlMs = ttlMs { body["ttl_ms"] = ttlMs }
        return try await request("POST", "/v1/services/\(Self.enc(service))/instances/\(Self.enc(instanceId))/heartbeat", body: body)
    }

    public func serviceDeregister(_ service: String, instanceId: String) async throws -> Any {
        try await request("DELETE", "/v1/services/\(Self.enc(service))/instances/\(Self.enc(instanceId))")
    }

    // MARK: - request core

    private func request(_ method: String, _ path: String, body: [String: Any]? = nil) async throws -> Any {
        guard let url = URL(string: baseURL + path) else {
            throw URLError(.badURL)
        }
        var req = URLRequest(url: url)
        req.httpMethod = method
        if let timeout = requestTimeout {
            req.timeoutInterval = timeout
        }
        if let body = body {
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
            req.httpBody = try JSONSerialization.data(withJSONObject: body, options: [])
        }

        let (data, response) = try await perform(req)
        let httpStatus = (response as? HTTPURLResponse)?.statusCode ?? 0
        let parsed: Any? = data.isEmpty
            ? nil
            : try? JSONSerialization.jsonObject(with: data, options: [.fragmentsAllowed])

        if httpStatus >= 300 {
            throw FiduciaError(status: httpStatus, body: parsed)
        }
        return parsed ?? NSNull()
    }

    /// Uses `URLSession.data(for:)` where available (macOS 12+/iOS 15+) and falls
    /// back to a `dataTask` continuation on older iOS so the async surface still
    /// works down to the package's iOS 13 floor.
    private func perform(_ request: URLRequest) async throws -> (Data, URLResponse) {
        if #available(macOS 12.0, iOS 15.0, tvOS 15.0, watchOS 8.0, *) {
            return try await session.data(for: request)
        } else {
            return try await withCheckedThrowingContinuation { continuation in
                let task = session.dataTask(with: request) { data, response, error in
                    if let error = error {
                        continuation.resume(throwing: error)
                    } else if let data = data, let response = response {
                        continuation.resume(returning: (data, response))
                    } else {
                        continuation.resume(throwing: URLError(.badServerResponse))
                    }
                }
                task.resume()
            }
        }
    }

    // RFC 3986 unreserved characters — safe in both path segments and query values.
    private static let encAllowed = CharacterSet(
        charactersIn: "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~"
    )

    private static func enc(_ value: String) -> String {
        value.addingPercentEncoding(withAllowedCharacters: encAllowed) ?? value
    }
}
